workflow/util/util.go

```diff
@@ -1049,16 +1058,18 @@ func GetTemplateFromNode(node wfv1.NodeStatus) string {
 
 func getNodeIDsToReset(restartSuccessful bool, nodeFieldSelector string, nodes wfv1.Nodes) (map[string]bool, error) {
 	nodeIDsToReset := make(map[string]bool)
-	if !restartSuccessful || len(nodeFieldSelector) == 0 {
+	if len(nodeFieldSelector) == 0 {
 		return nodeIDsToReset, nil
 	}
-
 	selector, err := fields.ParseSelector(nodeFieldSelector)
 	if err != nil {
 		return nil, err
 	} else {
 		for _, node := range nodes {
 			if SelectorMatchesNode(selector, node) {
+				if !restartSuccessful && node.Phase == wfv1.NodeSucceeded {
```

Comment:
I mentioned this [before](https://github.com/argoproj/argo-workflows/pull/12553#discussion_r1461432118), but the latter condition would be better as part of the selector if possible

Reply:
Hello. Here, I want to confirm whether restartSuccessful and selector have the highest priority of selector. If the selector selector selects a successful node but does not specify restartSuccessful, whether to re-execute the successful node. In this case, the restartSuccessful parameter doesn't feel very meaningful.

Reply:
The diagram looks correct to me

Reply:
> The diagram looks correct to me

However, the above diagram may be a bit in conflict with the previous logic, which may cause problems with the previous e2e test. For example, `TestFormulateRetryWorkflow/Nested_DAG_with_Non-group_Node_Selected`. Do you think it's better to fix only the previous red logic in the figure below and keep the previous blue logic?

![image](https://github.com/argoproj/argo-workflows/assets/8521943/a3dcf6bd-323f-48b7-a109-51ca43f43d96)

Reply:
Ahhhh -- based on the issue and your fix I strongly suspected there was something deeper incorrect with the previous logic. I actually [asked in the contributors channel](https://cloud-native.slack.com/archives/C0510EUH90V/p1707434418760689) if someone more familiar with the retry logic could check this PR because of that suspicion. 

> Do you think it's better to fix only the previous red logic in the figure below and keep the previous blue logic?

Ah is that what you were doing with your initial/current fix?
I honestly think we should change it all to work as expected (the black boxes) since the previous logic feels very confusing (and has confused users many a time before).

Big thanks for drawing the diagrams here, those are super helpful! We may want to add a similar `mermaid` flowchart of this to the docs as well

Reply:
If we rewrite this, we may want to release it as a breaking change similar to `retryStrategy` fixes from https://github.com/argoproj/argo-workflows/pull/11005

(different retries, but similar concept that they both behaved unexpectedly)

Reply:
> If we rewrite this, we may want to release it as a breaking change similar to `retryStrategy` fixes from #11005
> 
> (different retries, but similar concept that they both behaved unexpectedly)

By the way, do you have any plans to rewrite this part of the logic? Right now, our business needs to be able to retry a single error node. The changes to this issue are to minimize the impact on existing API logic and to provide the ability to retry a single faulty node. Of course, a broader refactoring might provide a more elegant solution. Do you know if the official team has any plans in this regard? 😃

Reply:
> Of course, a broader refactoring might provide a more elegant solution. Do you know if the official team has any plans in this regard? 😃

Should be happening shortly per [below](https://github.com/argoproj/argo-workflows/pull/12553#issuecomment-2404005416)